be/src/util/bit_packing_simd.h

```diff
@@ -0,0 +1,451 @@
+                            uint8_t* __restrict__ out);
+
+template <>
+inline void unpack<uint16_t>(int bit_width, const uint8_t* __restrict__ in, int64_t in_bytes, int64_t num_values,
+                             uint16_t* __restrict__ out);
+
+template <>
+inline void unpack<uint32_t>(int bit_width, const uint8_t* __restrict__ in, int64_t in_bytes, int64_t num_values,
+                             uint32_t* __restrict__ out);
+
+#if defined(__AVX2__) && defined(__BMI2__)
+// num_values number of uint16_t values with bit_width in
+//  [1, 7] range.
+static inline void unpack1to7(const uint8_t* __restrict__ in, uint16_t* __restrict__ out, int64_t num_values,
+                              int bit_width) {
+    uint64_t pdepMask = kPdepMask8[bit_width];
+    auto writeEndOffset = out + num_values;
+
+    while (out + 8 <= writeEndOffset) {
+        uint64_t val = *reinterpret_cast<const uint64_t*>(in);
+        uint64_t intermediateVal = _pdep_u64(val, pdepMask);
+
+        __m128i result = _mm_cvtepu8_epi16(_mm_loadl_epi64((reinterpret_cast<const __m128i*>(&intermediateVal))));
+        _mm_storeu_si128(reinterpret_cast<__m128i*>(out), result);
+
+        in += bit_width;
+        out += 8;
+    }
+}
+
+static inline void unpack8(const uint8_t* __restrict__ in, uint16_t* __restrict__ out, int64_t num_values) {
+    auto writeEndOffset = out + num_values;
+
+    while (out + 8 <= writeEndOffset) {
+        uint64_t value = *(reinterpret_cast<const uint64_t*>(in));
+
+        __m128i result = _mm_cvtepu8_epi16(_mm_loadl_epi64((const __m128i*)&value));
+        _mm_storeu_si128(reinterpret_cast<__m128i*>(out), result);
+
+        in += 8;
+        out += 8;
+    }
+}
+
+// Unpack num_values number of uint16_t values with bit_width in [9, 15] range.
+static inline void unpack9to15(const uint8_t* __restrict__ in, uint16_t* __restrict__ out, int64_t num_values,
+                               int bit_width) {
+    uint64_t pdepMask = kPdepMask16[bit_width];
+
+    uint8_t shift1 = bit_width * 4;
+    uint8_t shift2 = 64 - shift1;
+
+    auto writeEndOffset = out + num_values;
+
+    // Process bit_width bytes (8 values) a time.
+    while (out + 8 <= writeEndOffset) {
+        uint64_t value_0 = *(reinterpret_cast<const uint64_t*>(in));
+        uint64_t value_1 = *(reinterpret_cast<const uint64_t*>(in + 8));
+
+        *reinterpret_cast<uint64_t*>(out) = _pdep_u64(value_0, pdepMask);
+        *reinterpret_cast<uint64_t*>(out + 4) = _pdep_u64((value_0 >> shift1) | (value_1 << shift2), pdepMask);
+
+        in += bit_width;
+        out += 8;
+    }
+}
+
+// Unpack num_values number of uint16_t values with bit_width = 16
+static inline void unpack16(const uint8_t* __restrict__ in, uint16_t* __restrict__ out, int64_t num_values) {
+    uint64_t numBytes = num_values * 2;
+    std::memcpy(out, in, numBytes);
+
+    in += numBytes;
+    out += numBytes;
```

Comment:
out += num_values?

Reply:
Good catch, i will fix it, i think because we don't use the `out` result, so there is no effect on final result, but it's really a bug.